packages/instrumentor/plugins/functionHooks.ts

```diff
@@ -0,0 +1,234 @@
+
+function targetPath(path: NodePath<babel.Node>): string {
+	return path.getAncestry().reduce((acc: string, p: NodePath<babel.Node>) => {
+		if ("id" in p.node && babel.isIdentifier(p.node.id)) {
+			return addElementToPath(p.node.id.name, acc);
+		}
+		if ("key" in p.node) {
+			if (babel.isIdentifier(p.node.key)) {
+				return addElementToPath(p.node.key.name, acc);
+			} else if (babel.isStringLiteral(p.node.key)) {
+				return addElementToPath(p.node.key.value, acc);
+			}
+		}
+
+		if (babel.isAssignmentExpression(p.node)) {
+			return addElementToPath(generate(p.node.left).code, acc);
+		}
+
+		return acc;
+	}, "");
+}
+
+function addElementToPath(element: string, path: string): string {
+	const separator = path !== "" ? "." : "";
+	return element + separator + path;
+}
+
+function applyHooks(
+	functionName: string,
+	functionNode: babel.Function,
+	hooks: MatchingHooksResult
+): boolean {
+	if (!babel.isBlockStatement(functionNode.body)) {
+		functionNode.body = types.blockStatement([
+			types.returnStatement(functionNode.body),
+		]);
+	}
+
+	const origFuncName = functionName + "_original";
+	if (
+		hooks[HookType.Replace].length !== 0 ||
+		hooks[HookType.After].length !== 0
+	) {
+		functionNode.body = types.blockStatement([
+			createInternalFunctionFromBody(
+				origFuncName,
+				//TODO check this
+				functionNode.params as Array<babel.Identifier>,
+				functionNode.body
+			),
+		]);
+	}
+
+	if (hooks[HookType.Replace].length !== 0) {
+		functionNode.body.body.push(
+			types.returnStatement(
+				callHookExpression(
+					hooks[HookType.Replace][0],
+					functionNode.params as Array<babel.Identifier>,
+					[types.identifier(origFuncName)]
+				)
+			)
+		);
+	}
+
+	if (hooks[HookType.After].length !== 0) {
+		const retVal = types.identifier(origFuncName + "_result");
+		const origCal = callOriginalFunctionExpression(
+			origFuncName,
+			functionNode.params as Array<babel.Identifier>
+		);
+
+		if (hooks[HookType.After][0].async) {
+			let thenChainCallExpr = origCal;
+			for (const afterHook of hooks[HookType.After]) {
+				thenChainCallExpr = types.callExpression(
+					types.memberExpression(thenChainCallExpr, types.identifier("then")),
+					[
+						asyncHookThenExpression(
+							afterHook,
+							functionNode.params as Array<babel.Identifier>,
+							retVal
+						),
+					]
+				);
+			}
+			functionNode.body.body.push(types.returnStatement(thenChainCallExpr));
+		} else {
+			functionNode.body.body.push(
+				types.variableDeclaration("const", [
+					types.variableDeclarator(retVal, origCal),
+				])
+			);
+			for (const afterHook of hooks[HookType.After]) {
+				functionNode.body.body.push(
+					types.expressionStatement(
+						callHookExpression(
+							afterHook,
+							functionNode.params as Array<babel.Identifier>,
+							[retVal]
+						)
+					)
+				);
+			}
+			functionNode.body.body.push(types.returnStatement(retVal));
+		}
+	}
+
+	for (const beforeHook of hooks[HookType.Before].reverse()) {
```

Comment:
For symmetry to the other cases: Add `if (hooks[HookType.Befor].length !== 0) `?

Reply:
I refactored this code and split it into multiple functions.